### DISCUSSION TITLE
Is it possible to constrain repeated use of a TypeVar in function arguments to the same concrete type?

### DISCUSSION THREAD

**Discussion by CobaltCause** [microsoft/pyright#5476] [Q&A] [answered]
For example, I want some code like

```python
from typing import TypeVar

T = TypeVar("T")

def foo(x: T, y: T) -> None:
    pass

foo("bar", 4)
```

to fail to typecheck because `"bar"` and `4` are not the same type. (FWIW, the above code also passes mypy, so I'm guessing this is intended behavior and I'm missing some other way to write this so that it fails instead.)

---

To avoid the XY Problem, my actual situation is more like this:

```python
from abc import ABC
from asyncio import Queue
from typing import TypeVar

class SomeBaseClass(ABC):
    # define some abstract static methods

T = TypeVar("T", bound=SomeBaseClass)

def foo(queue: Queue[T]):
    # get the concrete type of T and call some abstract static methods from SomeBaseClass on it
```

but that seems to be impossible, hence the above.

**Comment by erictraut** [comment, marked as answer]
The constraint solver (the logic within a type checker that's responsible for "solving" a type variable) will do its best to find a type that satisfies all of the constraints. I your top example, the constraints for `T` can be satisfied with the type `str | int`, so that's considered a valid solution.

If you want to constrain a TypeVar to a particular list of possible types, you can use a constrained TypeVar, like this:

```python
T = TypeVar("T", int, str)
```

There is no way in the Python type system to constrain a TypeVar to a particular (arbitrary) concrete type because it's always possible for the constraint solver to widen the type with a union.

**Reply by CobaltCause** [reply]
Ohhh, that makes sense, thank you! The main piece I was missing was about how it *is* the same type, it's just been widened to a union.
